fix: User should be logged out when API sends 401 Unauthorized error

chavda-bhavik opened the pull request:
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
fix the issue of `user is not logging out when API sends 401 status code`

- **What is the current behavior?** (You can also link to an open issue here)
Regarding the [issue](https://github.com/novuhq/novu/issues/918), when API sends 401 (Unauthorized) error, nothing happens to UI.

- **What is the new behavior (if this is a feature change)?**
User redirects to `/auth/login` when API sends 401 (Unauthorized) error.

- **Other information**:
@davidsoderberg I was able to fix the issue by adding the `Axios` interceptor. But I'm facing one problem, the user gets redirected to the `/auth/login` page multiple times because of numerous API calls running simultaneously. Here are the steps to face that behavior, happy to hear any suggestions to solve it.
   1. Do login
   2. change `auth_token` in `localStorage`.
   3. refresh the page.

dcbartlett commented on apps/web/src/api/axios.client.ts:
```diff
@@ -0,0 +1,18 @@
+      localStorage.removeItem('auth_token');
+      localStorage.removeItem('widget_user_auth_token');
+      window.location.href = '/auth/login';
+    }
```
This makes an assumption that all unauthorized responses should also de-authenticate the user.  I feel this is not correct and should not be implemented.

scopsy replied:
Tend to aggree here, there was a discussion on Discord about that aswell.

scopsy commented:
@chavda-bhavik I think we can actually add an expiration check at the `RequiredAuth` component at `App.tsx` and logout there once the token is expired. What do you think?

chavda-bhavik commented:
Yes, @scopsy, that sounds good approach, I will update that way. Thank you.

scopsy commented:
@chavda-bhavik I've added the explanation here: https://docs.novu.co/community/run-locally#web

To run the API and WS for the tests use:

```shell
npm run start:e2e:api
npm run start:ws:test
```

chavda-bhavik commented:
**Here is what I have done**
- Reverted `Axios` interceptors which I added previously
- Added token expiry check in the `RequiredAuth` function in `App.tsx`
- Added test named `should logout user when auth token is expired`

**I've one doubt**
- Generated token has an expiry of 30 days `iat - exp = 30 days` but API starts sending 401 when the token is one day old

**I found one issue in tests**
- Test named `should redirect to the dashboard page when a token exists in query` in `auth.spec.ts` is failing
  - It tests that when the token is passed in `/auth/login?token=` the user should redirect to the `/templates` page, but the user is redirecting to the `/auth/application` page

davidsoderberg commented:
@chavda-bhavik yeah that test fails localy sometimes but all test passed in Github so it is fine :)

davidsoderberg commented:
```
I've one doubt

Generated token has an expiry of 30 days iat - exp = 30 days but API starts sending 401 when the token is one day old
```
Hm can this be a bug @scopsy ?

davidsoderberg reviewed (approved):
Looks good to me 🥳

scopsy commented on apps/web/cypress/tests/auth.spec.ts:
```diff
@@ -75,4 +75,26 @@ describe('User Sign-up and Login', function () {
+      todaysDate.setDate(todaysDate.getDate() + 30); // iat - exp = 30 days
+      cy.clock(todaysDate);
+
+      cy.visit('/templates');
```
Maybe worth adding here an actual assertion? For example to see that the new page is a a login page and the token on localstorage was cleaned. Wdyt?

chavda-bhavik replied:
Yes, it will be good, I will add a check for it.

scopsy commented on apps/web/cypress/tests/auth.spec.ts:
```diff
@@ -75,4 +75,26 @@ describe('User Sign-up and Login', function () {
+      cy.seed();
+    });
+
+    it('should logout user when auth token is expired', function () {
```
awesome that you did the test here! Nice job

chavda-bhavik replied:
Thank you.

scopsy commented:
> ```
> I've one doubt
> 
> Generated token has an expiry of 30 days iat - exp = 30 days but API starts sending 401 when the token is one day old
> ```
> 
> Hm can this be a bug @scopsy ?

This is probably the token from the notification center that has a smaller expiry time. So it is ok in this case

chavda-bhavik commented:
@scopsy Should I update the `should logout user when auth token is expired` test case to test token expiry after 1 day instead of 30 days? 

Regarding,
https://github.com/novuhq/novu/pull/935#issuecomment-1206156294

scopsy commented on apps/web/src/App.tsx:
```diff
@@ -194,6 +194,11 @@ function RequiredAuth({ children }: any) {
 
+  // Logout if token.exp is in the past (expired)
+  if (payload && payload.exp && payload.exp <= Date.now() / 1000) {
+    logout();
```
Also, a small suggestion, worth returning here so the rest of the function won't execute it may prevent a flicker of loading the current page.

chavda-bhavik replied:
Okay

scopsy commented:
@chavda-bhavik no it should be 30 days :)
